### ISSUE TITLE
Can I buy you Galaxy Buds Plus?

### ISSUE THREAD

**Issue by koderian** [ThePBone/GalaxyBudsClient#6]
Can I buy you Galaxy Buds Plus? Will you use the Buds Plus so the app will support it?

**Comment by ThePBone** [comment]
Wow, yeah of course. I'd definitely work on adding Buds+ plus support if you are considering donating a pair to me. :)

It just depends on how you want to send them to me: send money via Paypal and I'll buy them myself or you ship them directly to me.

**Comment by koderian** [comment]
Ok! cool. I'll see what I can do

**Comment by asherperlmuter** [comment]
@koderian make a post on the galaxybuds subreddit so we can all chip in

**Comment by koderian** [comment]
Got it! What would I say? Not gonna lie, this app is VERY useful and fast

**Comment by postacik** [comment]
Hi, I've just bought a pair of Galaxy Buds+ and I'm a developer and I want to contribute to this project if you can tell me where to start.

**Comment by ThePBone** [comment]
Well, the big problem with the Buds+ is they use a different SPP protocol which makes it impossible for me to provide you with the exact steps on how to do that.

I realized that after another developer messaged me as well and asked if I could help him to implement Buds+ support by exchanging Bluetooth dumps. Since the SPP implementation of the Buds is fundamentally different from the Buds+, I dropped the idea to do "remote-debugging" with him as it would have been very time-consuming to poke around in some raw binary data when I can't even verify its validity. 

You will need to reverse-engineer their RFCOMM/SPP protocol from scratch by intercepting raw incoming binary data from one of the Bluetooth channels/services (on Windows or Android).
Once you have collected enough binary data you need to analyze it using a HEX-editor. You need to find out where the actual SPP messages start and end, their size, type, id, and most importantly the algorithm used to validate and generate checksums from their actual payloads (for example SHA16, CRC16, ...). After that, you'd need to go one level deeper and start reversing the actual payloads contained in those messages.

TL;DR: Unfortunately I won't be a big help for adding Buds+ support since I don't own them.

**Comment by postacik** [comment]
Thanks for the deeeeep information. What's the best tool to capture raw bluetooth communication data?

**Comment by ThePBone** [comment]
There is no suitable one. I wrote my own C# development tool (included with
the manager app, it won't work with the Buds+ obviously). But writing a
small Python script to receive/send from a RFCOMM socket should be enough
too.

**Comment by superbonaci** [comment]
> Wow, yeah of course. I'd definitely work on adding Buds+ plus support if you are considering donating a pair to me. :)
> 
> It just depends on how you want to send them to me: send money via Paypal and I'll buy them myself or you ship them directly to me.

I think if you add them to your Amazon's wishlist anybody can gift that to you.

**Comment by superbonaci** [comment]
> Thanks for the deeeeep information. What's the best tool to capture raw bluetooth communication data?

I use HCI snoop log from Android. If you have rooted LineageOS it's easy peasy.

**Comment by postacik** [comment]
> > Thanks for the deeeeep information. What's the best tool to capture raw bluetooth communication data?
> 
> I use HCI Snoop Log from Android. If you have rooted LineageOS it's easy peasy to copy the log file from data partition. Then use WireShark to view raw data.

Unfortunately I use a non rooted Samsung phone.

I want to ask another question about reverse engineering. On my phone I have the Galaxy Buds+ app installed which I think I can  copy the apk to my computer and have a look at the Java code for bluetooth communications.

Would it be of any help if I could extract the Java data structures from that package?

**Comment by superbonaci** [comment]
@postacik search under Developer options.

**Comment by postacik** [comment]
Thanks for the tip. I've worked with apk tools and Java decompilers before. I just wondered if @ThePBone ever tried to figure out the communication protocol by looking at the source code.

**Comment by ThePBone** [comment]
> Thanks for the tip. I've worked with apk tools and Java decompilers before. I just wondered if @ThePBone ever tried to figure out the communication protocol by looking at the source code.

Yes, otherwise I wouldn't have been able to get ahold of all their message ids and parameter-information in there. I've also had a quick look at the Buds+ plugin APK disassembly to check the difference without actually having a pair. 

I recommend you to use <https://github.com/skylot/jadx> since it is capable of disassembling Smali bytecode to more or less accurate Java code (no idea if APK Easy Tool has a dex-to-java disassembler as well, but IIRC I don't think it has one unless you want to read Smali code).

You might need to enable "Preferences > Show inconsistent code" in jadx since some parts of the SPP message handler are lost (be careful though you might end up with invalid code)

**Comment by ThePBone** [comment]
@postacik I've also uploaded my dev-tool (C#) modified for the Buds+ the other week for another dev. 

I already replaced all messaged IDs with the ones for the Buds+ and also updated the SDP UUID so it should connect to the right Bluetooth interface. I don't expect it to work on the first try (crashes are expected) but you could use it as a starting point. You can find a brief description in the README. 

<https://github.com/ThePBone/GalaxyBudsDevTool>

**Comment by ThePBone** [comment]
> There are also good discounts on eBay.

That's actually a really good idea, I can't believe I haven't thought of that before.
I'll see if I can get a good deal, maybe a used/refurbished pair if they aren't too expensive.

**Comment by superbonaci** [comment]
Glad you found them.

**Comment by ThePBone** [comment]
This update adds support for the Buds+:<https://github.com/ThePBone/GalaxyBudsClient/releases/tag/2.0.0>

**Comment by postacik** [comment]
Great news. Thanks :)

**Comment by superbonaci** [comment]
Good job.
